### ISSUE TITLE
wakati

### ISSUE THREAD

**Issue by kuniotoku** [kuniotoku/chatGPT#3]

**Comment by kuniotoku** [comment]
import pandas as pd
from collections import Counter
import nltk
nltk.download('stopwords')
from nltk.corpus import stopwords

# DataFrameの例
df = pd.DataFrame({
    'ID': [1, 1, 2, 2, 2],
    'wakati': ['the quick brown fox', 'jumps over the lazy dog', 
               'the quick brown fox', 'the quick brown fox', 'over the lazy dog']
})

# ストップワードを取得
stop_words = set(stopwords.words('english'))

# 同じIDごとに単語を集計する関数
def count_words(group):
    # 同じID内の全ての単語を結合し、空白で分割する
    words = ' '.join(group['wakati']).split()
    # ストップワードを除外する
    words = [word for word in words if word.lower() not in stop_words]
    # 単語の出現回数をカウントする
    word_counts = Counter(words)
    # 最も頻出する単語を返す
    return word_counts.most_common(1)[0][0]

# 同じIDごとに最も頻出する単語を集計する
most_common_words = df.groupby('ID').apply(count_words)

# 結果を表示する
print(most_common_words)

**Comment by kuniotoku** [comment]
import pandas as pd
from collections import Counter
from sklearn.feature_extraction.text import ENGLISH_STOP_WORDS

# DataFrameの例
df = pd.DataFrame({
    'ID': [1, 1, 2, 2, 2],
    'wakati': ['the quick brown fox', 'jumps over the lazy dog', 
               'the quick brown fox', 'the quick brown fox', 'over the lazy dog']
})

# ストップワードを取得
stop_words = ENGLISH_STOP_WORDS

# 同じIDごとに単語を集計する関数
def count_words(group):
    # 同じID内の全ての単語を結合し、空白で分割する
    words = ' '.join(group['wakati']).split()
    # ストップワードを除外する
    words = [word for word in words if word.lower() not in stop_words]
    # 単語の出現回数をカウントする
    word_counts = Counter(words)
    # 最も頻出する単語を返す
    return word_counts.most_common(1)[0][0]

# 同じIDごとに最も頻出する単語を集計する
most_common_words = df.groupby('ID').apply(count_words)

# 結果を表示する
print(most_common_words)

**Comment by kuniotoku** [comment]
import pandas as pd
import networkx as nx
import matplotlib.pyplot as plt

df = pd.read_csv("data.csv")

from itertools import combinations
from collections import defaultdict

cooccur_dict = defaultdict(int)

for _, group in df.groupby("ID"):
    words = set()
    for w in group["wakati"]:
        words |= set(w.split())
    for pair in combinations(words, 2):
        cooccur_dict[pair] += 1

top_n = 20

G = nx.Graph()

for (w1, w2), count in sorted(cooccur_dict.items(), key=lambda x: -x[1])[:top_n]:
    G.add_edge(w1, w2, weight=count)

pos = nx.spring_layout(G, k=0.2)
nx.draw(G, pos, node_size=2000, font_size=16, with_labels=True, font_weight="bold")
labels = nx.get_edge_attributes(G, "weight")
nx.draw_networkx_edge_labels(G, pos, edge_labels=labels, font_size=14)
plt.show()

**Comment by kuniotoku** [comment]
# 同じIDごとに単語を集計する関数
def count_words(group, n):
    # 同じID内の全ての単語を結合し、空白で分割する
    words = ' '.join(group['wakati']).split()
    # ストップワードを除外する
    words = [word for word in words if word.lower() not in stop_words]
    # 単語の出現回数をカウントする
    word_counts = Counter(words)
    # 上位n個の単語をリストで返す
    return [word for word, count in word_counts.most_common(n)]

# 同じIDごとに上位n個の単語を集計する
n = 2  # 上位n個
most_common_words = df.groupby('ID').apply(count_words, n=n)

# 結果を表示する
id_col = 'ID'
word_cols = [f"word_{i+1}" for i in range(n)] # カラム名を指定する
count_cols = [f"count_{i+1}" for i in range(n)] # カラム名を指定する

# IDとn個の単語カラム、n個の出現回数カラムからなるデータフレームを作成する
id_groups = most_common_words.groupby(id_col)
result = pd.concat([id_groups[0].apply(pd.Series) for i, id_groups in id_groups], axis=1)
result.columns = [id_col] + [f"{col}_{i+1}" for i, id_groups in id_groups for col in [word_cols, count_cols]]

print(result)

**Comment by kuniotoku** [comment]
import pandas as pd
from collections import Counter
from sklearn.feature_extraction.text import ENGLISH_STOP_WORDS

# DataFrameの例
df = pd.DataFrame({
    'ID': [1, 1, 2, 2, 2],
    'wakati': ['the quick brown fox', 'jumps over the lazy dog', 
               'the quick brown fox', 'the quick brown fox', 'over the lazy dog']
})

# ストップワードを取得
stop_words = ENGLISH_STOP_WORDS

# 同じIDごとに単語を集計する関数
def count_words(group, n):
    # 同じID内の全ての単語を結合し、空白で分割する
    words = ' '.join(group['wakati']).split()
    # ストップワードを除外する
    words = [word for word in words if word.lower() not in stop_words]
    # 単語の出現回数をカウントする
    word_counts = Counter(words)
    # 上位n個の単語を返す
    return dict(word_counts.most_common(n))

# 同じIDごとに上位n個の単語を集計する
n = 2  # 上位n個
most_common_words = df.groupby('ID').apply(count_words, n=n).apply(pd.Series)

# カラム名を変更する
column_names = {(i, j): f'word_{j+1}_count' for i, j in most_common_words.columns}
most_common_words = most_common_words.rename(columns=column_names)

# 結果を表示する
print(most_common_words)
